input-explainer.md

```diff
@@ -280,6 +292,91 @@ When the canvas receives a `pointerdown` event an `XRInputSource` is created wit
+  - Inline sessions MUST only expose `id`s of `'unknown'`.
+  - Otherwise the `id` should be a lower-case string that describes the physical input source.
+    - For most devices this SHOULD be of the format `<vendor>-<product-id>`. For example: `oculus-touch`. UAs SHOULD make an effort to align on the strings that are returned for any given device.
+    - It should not include an indication of the handedness of the input source, as that should be implied by the `handedness` attribute.
+
+All other attributes behave as described in the [Gamepad](https://w3c.github.io/gamepad/) specification.
+
+The WebXR Device API introduces a new standard controller layout indicated by the `mapping` value of `xr-standard`. (Additional mapping variants may be added in the future if necessary.) This defines a specific layout for the inputs most commonly found on XR controller devices today. The following table describes the buttons/axes and their physical locations:
+
+| Button/Axis  | Location                          |
+| ------------ | --------------------------------- |
+| buttons[0]   | Primary trigger                   |
+| buttons[1]   | Primary Touchpad/Joystick click   |
+| buttons[2]   | Grip/Secondary trigger            |
+| buttons[3]   | Secondary Touchpad/Joystick click |
+| axes[0]      | Primary Touchpad/Joystick X       |
+| axes[1]      | Primary Touchpad/Joystick Y       |
+| axes[2]      | Secondary Touchpad/Joystick X     |
+| axes[3]      | Secondary Touchpad/Joystick Y     |
+
+Additional device-specific inputs may be exposed after these reserved indices, but devices that lack one of the canonical inputs must still preserve their place in the array. If a device has both a touchpad and a joystick the UA should designate one of them to be the primary axis-based input and expose the other at axes[2] and axes[3] with an associated button at button[3].
+
+```js
+function onXRFrame(timestamp, frame) {
+  let inputSource = primaryInputSource;
+
+  // Check to see if the input source has buttons/axes and is using the standard
+  // mapping.
+  if (inputSource && inputSource.gamepad &&
+      inputSource.gamepad.mapping == 'xr-standard') {
+    let gamepad = inputSource.gamepad;
+    
+    // Use joystick or touchpad values for movement.
+    MoveUser(gamepad.axes[0], gamepad.axes[1]);
+
+    // Use the trigger or joystick/touchpad click for painting.
+    if (gamepad.buttons[0].pressed || gamepad.buttons[1].pressed) {
+      EmitPaint();
+    }
+    
+    // etc.
+  }
+
+  // Do the rest of typical frame processing...
+}
+```
+
+If the application includes interactions that require user activation (such as starting media playback), the application can listen to the `XRInputSource`s `select` events, which fire for every pressed and released button on the controller. When triggered by a controller input, the `XRInputSourceEvent` will include a `buttonIndex` other than `-1` to indicate which button on the gamepad triggered the event.
```

Comment:
How do we plan to rationalize with the design work happening around events in the gamepad spec? If we allow those events to fire, we're somewhat breaking the mental model that all XR input fires one event type.  We'd also need to deal with the fact that those events wouldn't be aligned to XRFrame timing necessarily.  Also, we should find out the timeframe in which the design might actually be available to be used? https://github.com/w3c/gamepad/pull/15

Reply:
The lack of XRFrame alignment is the biggest issue I would have with trying to use the gamepad events. I think we'd probably just state that gamepads not exposed by `navigator.getGamepads()` don't fire standard gamepad events?

Reply:
Should we be talking to them about including an optional XRFrame on their events?

Reply:
I don't know that I feel that's the right solution. Maybe I'm just a bit gunshy, though, because we previously tried to pile a bunch of VR stuff on top of Gamepads and that didn't work out very well.

Reply:
I'm ok to file an issue in the WebXR repo to track a unification plan with those proposed events.